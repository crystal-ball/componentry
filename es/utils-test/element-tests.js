/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { mount } from 'enzyme';
import dt from './dt';
/**
 * Library default component requirment test suite.
 *
 * IMPORTANT NOTES
 * - Enzyme `find()` will find **EVERY** instance of a prop, this makes it difficult
 *   to assert on find lengths, because it will be variable for components with
 *   HOCs which is why we use `toBeTruthy` to only check that some instance of the
 *   expected element attr exists
 */

export default (function (TestComponent) {
  var enzymeOptions = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

  /*
   * Test that any html tag or component can be passed through the 'as' prop to
   * override the parent element the component renders
   */
  test('should render as specified html element or component', function () {
    // Create a component to validate that the TestComponent returns.
    var TestAs = function TestAs() {
      return /*#__PURE__*/React.createElement("section", null);
    };

    var wrapper = mount( /*#__PURE__*/React.createElement(TestComponent, {
      as: TestAs
    }), enzymeOptions);
    expect(wrapper.find(TestAs).length).toBeTruthy();
  });
  /*
   * All components should spread passed props on the parent container for max
   * customization
   */

  test('should pass through classnames and attributes', function () {
    var wrapper = mount( /*#__PURE__*/React.createElement(TestComponent, {
      className: "test-custom",
      "data-test": "test-custom"
    }), enzymeOptions);
    expect(wrapper.find('.test-custom').length).toBeTruthy();
    expect(wrapper.find(dt('test-custom')).length).toBeTruthy();
  });
});