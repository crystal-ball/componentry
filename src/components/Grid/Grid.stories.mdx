import { ArgsTable, Meta, Story } from '@storybook/addon-docs'
import { BADGE } from '@geometricpanda/storybook-addon-badges'

import { Grid } from './Grid'

export function padNum(num) {
  return num > 9 ? num : `0${num}`
}

<Meta
  title='Components/Grid'
  component={Grid}
  parameters={{ badges: [BADGE.EXPERIMENTAL] }}
/>

# Grid

<Story name='Grid'>
  <Grid gap={2} backgroundColor='primary-200' className='rounded grid-cols-2'>
    {[1, 2].map((key) => (
      <div
        key={key}
        className='bg-primary-600 text-inverse font-mono font-bold px-3 py-2 rounded'
      >
        #{padNum(key)}
      </div>
    ))}
  </Grid>
</Story>

Grid is a layout component with shorthand properties `align`, and `justify` for common
alignment styles.

<ArgsTable of={Grid} />
