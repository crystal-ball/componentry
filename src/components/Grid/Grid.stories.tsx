import type { Meta, StoryObj } from '@storybook/react'

import { Grid } from './Grid'

const meta: Meta<typeof Grid> = {
  component: Grid,
}

export default meta
type Story = StoryObj<typeof Grid>

export const Primary: Story = {
  args: {
    backgroundColor: 'primary-200',
    className: 'rounded grid-cols-2',
    gap: 4,
    children: [1, 2, 3].map((num) => (
      <div
        key={num}
        className='bg-primary-600 text-inverse font-mono font-bold px-6 py-4 rounded'
      >
        #{num > 9 ? num : `0${num}`}
      </div>
    )),
  },
}
