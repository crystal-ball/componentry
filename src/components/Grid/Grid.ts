/* eslint-disable react/no-unused-prop-types */
import { type FC, forwardRef } from 'react'
import { type ComponentBaseProps } from '../../utils/base-types'
import { element } from '../../utils/element-creator'
import { useThemeProps } from '../Provider/Provider'

export interface GridPropsBase {
  /** Sets an `align-items` style */
  align?: 'start' | 'end' | 'center' | 'baseline' | 'stretch'
  /** Sets a `justify-items` style */
  justify?: 'start' | 'end' | 'center' | 'stretch'
}

export type GridProps = GridPropsBase & ComponentBaseProps<'div'>

/**
 * **[üìù Grid docs](https://componentry.design/docs/components/grid)**
 *
 * `Grid` provides consistent CSS grid layouts using a grid container with
 * utility props.
 * @example
 * ```tsx
 * <Grid gap={2}>
 *   ...
 * </Grid>
 * ```
 */
export const Grid: FC<GridProps> = (props) => {
  const { align, justify, ...rest } = {
    ...useThemeProps('Grid'),
    ...props,
  }

  return element({
    display: 'grid',
    alignItems: align,
    justifyItems: justify,
    ...rest,
  })
}
Grid.displayName = 'Grid'

/**
 * **[üìù Grid docs](https://componentry.design/docs/components/grid)**
 *
 * `Grid` provides a ref-able `Grid` element.
 * @example
 * ```tsx
 * <RefGrid ref={ref} gap={2}>
 *   ...
 * </RefGrid>
 * ```
 */
export const RefGrid = forwardRef<HTMLDivElement, GridProps>((props, ref) => {
  const { align, justify, ...rest } = {
    ...useThemeProps('Grid'),
    ...props,
  }

  return element({
    ref,
    display: 'grid',
    alignItems: align,
    justifyItems: justify,
    ...rest,
  })
})
RefGrid.displayName = 'RefGrid'
