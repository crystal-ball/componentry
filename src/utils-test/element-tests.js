/* eslint-disable import/no-extraneous-dependencies */
import React from 'react'
import { mount } from 'enzyme'

import dt from './dt'

/**
 * Library default component requirment test suite.
 *
 * IMPORTANT NOTES
 * - Enzyme `find()` will find **EVERY** instance of a prop, this makes it difficult
 *   to assert on find lengths, because it will be variable for components with
 *   HOCs which is why we use `toBeTruthy` to only check that some instance of the
 *   expected element attr exists
 */

export default (TestComponent, enzymeOptions = {}) => {
  /*
   * Test that any html tag or component can be passed through the 'as' prop to
   * override the parent element the component renders
   */
  test('should render as specified html element or component', () => {
    // Create a component to validate that the TestComponent returns.
    const TestAs = () => <section />
    const wrapper = mount(<TestComponent as={TestAs} />, enzymeOptions)
    expect(wrapper.find(TestAs).length).toBeTruthy()
  })

  /*
   * All components should spread passed props on the parent container for max
   * customization
   */
  test('should pass through classnames and attributes', () => {
    const wrapper = mount(
      <TestComponent className="test-custom" data-test="test-custom" />,
      enzymeOptions,
    )
    expect(wrapper.find('.test-custom').length).toBeTruthy()
    expect(wrapper.find(dt('test-custom')).length).toBeTruthy()
  })

  // TODO: should render passed children
}
