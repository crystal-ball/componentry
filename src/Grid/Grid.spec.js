import React from 'react'
import { render, screen } from '@testing-library/react'

import { elementTests } from '../test/element-tests'
import { Grid } from './Grid'

describe('<Grid />', () => {
  elementTests(Grid)

  it('When inline is not passed, then block className is rendered', () => {
    render(<Grid>Content</Grid>)

    expect(screen.getByText('Content')).toHaveClass('ðŸ…²-grid grid')
  })

  it('When inline is passed, then inline-block className is rendered', () => {
    render(<Grid inline>Content</Grid>)

    expect(screen.getByText('Content')).toHaveClass('ðŸ…²-grid inline-grid')
  })
})

// Snapshots
// ---------------------------------------------------------------------------
describe('<Grid /> snapshots', () => {
  it('renders correctly', () => {
    render(<Grid data-testid='grid'>Block content</Grid>)

    expect(screen.getByTestId('grid')).toMatchSnapshot()
  })
})
