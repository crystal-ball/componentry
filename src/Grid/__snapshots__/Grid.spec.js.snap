// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Grid /> snapshots renders correctly 1`] = `
<div
  class="grid"
  data-testid="grid"
>
  Block content
</div>
`;
